### PR TITLE
Add date+time support (JAVA-404)

### PR DESCRIPTION
**Pull request by olim7t**
This builds on @snazy 's contribution in #355 to add the following:
- full ISO8601 support to match exactly what's done by Joda Time on the C\* side (Java's built-in calendar is a bit different so we need a few hacks)
- optimization to create at most one `Calendar` instance per `DateWithoutTime`
- since the change will now break at compile time, I've renamed `getDateWithoutTime` to `getDate`
- actually I didn't mind relying on overflow to convert the unsigned int, so I've restored that instead of the code I had suggested

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/AbstractData.java [hunk 1]
```diff
@@ -139,13 +139,21 @@ public T setLong(String name, long v) {
     }
 
     public T setDate(int i, Date v) {
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/AbstractData.java]: Since getDate is being replaced with getTimestamp, should't we just get rid of setDate since there is no symmetric get for it?

File: driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+ */
+public final class DateWithoutTime {
+
+    private static final TimeZone UTC = TimeZone.getTimeZone("UTC");
+
+    private final long millisSinceEpoch;
+    private final int daysSinceEpoch;
+
+    // This gets initialized lazily if we ever need it. Once set, it is effectively immutable.
+    private volatile GregorianCalendar calendar;
```
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java:48]: why not make this static?

File: driver-core/src/test/java/com/datastax/driver/core/DateWithoutTimeTest.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+import java.util.concurrent.TimeUnit;
+
+import org.testng.annotations.Test;
+
+import static com.datastax.driver.core.Assertions.assertThat;
+import static com.datastax.driver.core.DateWithoutTime.fromDaysSinceEpoch;
+import static com.datastax.driver.core.DateWithoutTime.fromMillisSinceEpoch;
+import static com.datastax.driver.core.DateWithoutTime.fromYearMonthDay;
+
+public class DateWithoutTimeTest {
```
Comment (tolbertam) [inline driver-core/src/test/java/com/datastax/driver/core/DateWithoutTimeTest.java:27]: Very nice! Covers the corner cases with far in the past / in the future dates.

File: driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java (hunk 2 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java:48]: It can't be shared between instances.

File: driver-core/src/main/java/com/datastax/driver/core/AbstractData.java (hunk 1 above)
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/AbstractData.java]: I forgot to remove this during a refactoring.

File: driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java (hunk 2 above)
Comment (tolbertam) [inline driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java:48]: Ah, I got the getCalendar and isoCalendar() implementations mixed up, makes sense that you want a unique calendar per instance.

File: driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+     * @param year the year in ISO format ({@link DateWithoutTime this class's Javadoc}).
+     * @param month the month. It is 1-based (e.g. 1 for January) and not lenient.
+     * @param dayOfMonth the day of the month.
+     *
+     * @return the new instance.
+     *
+     * @throws IllegalArgumentException if the date is not in the range [-5877641-06-23; 5881580-07-11],
+     * or the month is not between 1 and 12, or the day is not between 1 and 31.
+     */
+    public static DateWithoutTime fromYearMonthDay(int year, int month, int dayOfMonth) {
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java:99]: `fromYearMonthDay(2015, 2, 31)` does not complain and yields `2015-03-03`, is that per design?

Comment (snazy) [inline driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java:99]: I guess a `Canendar.setLenient(false)` in `isoCalendar` should force it to complain on that date.

Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/DateWithoutTime.java:99]: > (in fact, the day is still lenient if <= 31, but this class makes no guarantee that this will remain true in the future)

Not allowing > 31 is necessary for the year hack to work, otherwise the user could provide (0,12,32) which is really (1,1,1), but we would shift the year by one because it's <= 0.

I saw no value in making the underlying calendar lenient, but in hindsight that's more confusing than useful, so I will follow Robert's suggestion.
